proposals/NNNN-hashable-enhancements.md

```diff
@@ -0,0 +1,925 @@
+adapt it to the special requirements of certain environments (such as
+network services), or to generally improve hashing performance.)
+
+## <a name="alternatives">Alternatives considered</a>
+
+### <a name="leave-hashable-alone">Leaving `Hashable` as is</a>
+
+One option that we considered is to expose `Hasher`, but to leave the
+`Hashable` protocol as is. Individual `Hashable` types would be able
+to choose whether or not to use it or to roll their own hash
+functions.
+
+We felt this was an unsatisfying approach, and we gave an extensive
+rationale for changing `Hashable` in the section on [The `hash(into:)`
+requirement](#hash-into).
+
+### <a name="protocol-refinement">Defining a protocol refinement</a>
+
+There have been several attempts ([1][h1], [2][h2], [3][h3]) to fix `Hashable`
+by creating a new protocol that refines it:
+
+[h1]: https://github.com/apple/swift/blob/swift-4.1-branch/validation-test/stdlib/HashingPrototype.swift
+[h2]: https://github.com/attaswift/SipHash
+[h3]: https://gist.github.com/regexident/1b8e84974da2243e5199e760508d2d25
+
+```swift
+protocol Hashable {
+  var hashValue: Int { get }
+}
+
+protocol Hashable2: Hashable {
+  func hash(into hasher: inout Hasher)
+}
+
+extension Hashable2 {
+  var hashValue: Int {
+    var hasher = Hasher()
+    hash(into: &hasher)
+    return hasher.finalize()
+  }
+}
+```
+
+While this is a useful approach for external hashing packages, we
+believe it to be unsuitable for the standard library. Adding a new
+protocol would add a significant new source of user confusion about
+hashing, and it would needlessly expand the standard library's API
+surface area.
+
+The new protocol would need to have a new name, but `Hashable` already
+has the perfect name for a protocol representing hashable things -- so
+we'd need to choose an imperfect name for the "better" protocol.
+
+While deprecating a protocol requirement is a significant change, we
+believe it to be less harmful overall than leaving `Hashable`
+unchanged, or trying to have two parallel protocols for the same thing.
+
+Additionally, adding a second protocol would lead to complications
+with `Hashable` synthesis. It's also unclear if `Set` and `Dictionary`
+would be able to consistently use `Hasher` for their primary hashing
+API. (These problems are not unsolvable, but they may involve adding
+special one-off compiler support for the new protocol. For example, we
+may want to automatically derive `Hashable2` conformance for all types
+that implement `Hashable`.)
+
+### <a name="generic-hasher">Making `hash(into:)` generic over a `Hasher` protocol</a>
+
+It would be nice to allow Swift programmers to define their own hash
+functions, and to plug them into any `Hashable` type:
+
+```swift
+protocol Hasher {
+  func append(bits: Int)
+  func append(bits: UInt)
+  /// etc.
+}
+protocol Hashable {
+  func hash<H: Hasher>(into hasher: inout H)
+}
+```
+
+However, we believe this would add an unnnecessary degree of freedom.
```

Comment:
I think "unnnecessary degree of freedom" will trigger outcry. Instead, the message is that this degree of generality would carry unjustifiable costs. That is, it requires losses that we believe to be unjustifiable relative to potential gain.

Reply:
Hm, how about this?

> However, we believe this would add a degree of generality whose costs are unjustifiable relative to their potential gain. We expect the ability to create custom hashers would rarely be exercised; in particular, we do not foresee a need for etc.

(I love how I accidentally used three nnns in 'unnnecessary'.)

Reply:
The single use case for custom hashers I can think of is in testing. However, mock hashers can easily be replaced by feeding the expected byte sequence to a regular hasher, and comparing finalized results.